File: frontend/src/api/npm/createUser.ts
import * as api from "./base";
import { User } from "./models";

export interface AuthOptions {
	type: string;
	secret: string;
}

export interface NewUser {
	name: string;
	nickname: string;
	email: string;
	isDisabled: boolean;
	auth: AuthOptions;
	capabilities: string[];
}

export async function createUser(
	data: NewUser,
	abortController?: AbortController,
): Promise<User> {
	const { result } = await api.post(
		{
			url: "/users",
			data,
		},
		abortController,
	);
	return result;
}

File: backend/internal/api/schema/create_user.go
package schema

import "fmt"

// CreateUser is the schema for incoming data validation
func CreateUser() string {
	return fmt.Sprintf(`
		{
			"type": "object",
			"additionalProperties": false,
			"required": [
				"name",
				"email",
				"is_disabled",
				"capabilities"
			],
			"properties": {
				"name": %s,
				"email": %s,
				"is_disabled": {
					"type": "boolean"
				},
				"auth": {
					"type": "object",
					"required": [
						"type",
						"secret"
					],
					"properties": {
						"type": {
							"type": "string",
							"pattern": "^local$"
						},
						"secret": %s
					}
				},
				"capabilities": %s
			}
		}
	`, stringMinMax(2, 50), stringMinMax(5, 150), stringMinMax(8, 255), capabilties())
}
